console: allow using arrow left/right, home/end keys and history mode

ngxson opened the pull request:
Fix https://github.com/ggml-org/llama.cpp/issues/17828

Fix https://github.com/ggml-org/llama.cpp/issues/17829

To test it, run `llama-cli`, write some text (without sending it) and use left/right arrows to edit them

Edit: Home/end keys and history support is allso added thanks to @ServeurpersoCom 

https://github.com/user-attachments/assets/afaa2399-21cb-4f45-adc1-97279c8c7302

Disclosure AI usage: `next_utf8_char_pos` is written by AI because I have no experience working with windows API. The rest is mostly written by me (minus some small auto-completion)

ngxson committed:
console: allow using arrow left/right to edit the line (with UTF-8 support)

ngxson committed:
console: allow using arrow left/right to edit the line (with UTF-8 support)

ServeurpersoCom committed:
console: fix arrow keys on Windows using private-use Unicode

ServeurpersoCom committed:
console: fix arrow keys on Windows using private-use Unicode

ServeurpersoCom committed:
console: add Home/End key support for Windows and Linux

ServeurpersoCom committed:
console: add Home/End key support for Windows and Linux

ServeurpersoCom committed:
console: add basic Up/Down history navigation

ServeurpersoCom committed:
console: add basic Up/Down history navigation

ngxson committed:
Merge pull request #53 from ServeurpersoCom/xsn/cli_arrow_left_right

Xsn/cli arrow left right

ngxson committed:
fix build

ServeurpersoCom committed:
console: remove unreachable wc == 0 check after VK switch

ServeurpersoCom committed:
console: add Ctrl+Left/Right word navigation

- Add KEY_CTRL_ARROW_LEFT and KEY_CTRL_ARROW_RIGHT codes
- Windows: detect CTRL modifier via dwControlKeyState
- Linux: parse ANSI sequences with modifier (1;5D/C)
- Implement move_word_left/right with space-skipping logic
- Refactor escape sequence parsing to accumulate params

ggerganov reviewed (approved):
Tested on Mac, works as expected.

- Would be nice to add support for "Delete" if possible
- During chat, when I type a new message and hit "Up + Down" it forgets what I wrote - could be improved.

ServeurpersoCom committed:
console: add Delete key support

- Windows: VK_DELETE detection
- Linux: ESC[3~ sequence parsing
- Forward character deletion with UTF-8 support

ServeurpersoCom committed:
console: implement bash-style history editing

- Edit any history line during UP/DOWN navigation, edits persist
- Pressing Enter appends edited version as new history entry
- Original line stay untouched in their positions

ngxson committed:
Merge remote-tracking branch 'pascal/console/terminal-improvement' into xsn/cli_arrow_left_right

ngxson committed:
clean up

ngxson committed:
better history impl

ngxson committed:
fix decode_utf8

ngxson commented:
@ggerganov could you test again to see if the 2 problems in your last comment are all resolved? Thanks!

ggerganov commented:
Yes, they are fixed.

Btw, the color of the first message is white and the next messages are green, is this expected?

<img width="501" height="228" alt="image" src="https://github.com/user-attachments/assets/58f1037f-39c4-48b8-bbf4-3df918bfc9fb" />

This is with:

```bash
llama-cli -m ../models/gpt-oss-20b/ggml-model-mxfp4.gguf -c 0
```
